catalog/controller/extension/payment/razorpay.php

```diff
@@ -94,59 +126,124 @@ private function get_order_creation_data($order_id)
         return $data;
     }
 
+    public function validate_non_recurring_products()
+    {
+        $nonRecurringProduct = array_filter($this->cart->getProducts(), function ($product) {
+            return array_filter($product, function ($value, $key) {
+                return $key == "recurring" && empty($value);
+            }, ARRAY_FILTER_USE_BOTH);
+        });
+
+        if (!empty($nonRecurringProduct)) {
+            $this->log->write("Cart has recurring product and non recurring product");
+            echo "<div class='alert alert-danger alert-dismissible'>You cannot have non-recurring and recurring product in your shopping cart </div>";
+            exit;
+        }
+    }
+
+    private function get_subscription_order_creation_data($order_id)
+    {
+        $this->load->model('extension/payment/razorpay');
+
+        $order = $this->model_checkout_order->getOrder($order_id);
+        $recurringPlanData = $this->cart->getProducts()[0]["recurring"];
+        $productId = $this->cart->getProducts()[0]['product_id'];
+
+        $planData = $this->model_extension_payment_razorpay->getPlanByRecurringIdAndFrequencyAndProductId($recurringPlanData['recurring_id'], $recurringPlanData['frequency'], $productId);
```

Comment:
list of plan details from recurring table. can you map recurring table and plan id